src/switchroot/ostree-prepare-root.c

```diff
@@ -684,12 +715,60 @@ main (int argc, char *argv[])
   if (rmdir (TMP_SYSROOT) < 0)
     err (EXIT_FAILURE, "couldn't remove temporary sysroot %s", TMP_SYSROOT);
 
-  /* Now that we've set up all the mount points, if configured we remount the physical
-   * rootfs as read-only; what is visibly mutable to the OS by default is just /etc and /var.
-   * But ostree knows how to mount /boot and /sysroot read-write to perform operations.
-   */
-  if (sysroot_readonly)
+  if (sysroot_invisible)
+    {
+      /* Keep a living sysroot in a private mount namespace,
+       * and unmount sysroot in the root mount namespace to make it invisible.
+       */
+      const char *sysroot_ns = OTCORE_RUN_OSTREE_PRIVATE "/sysroot-ns";
+      glnx_autofd int ns_fd = open (sysroot_ns, O_WRONLY | O_CREAT, 0);
+      if (ns_fd < 0)
+        err (EXIT_FAILURE, "failed to create %s", sysroot_ns);
+
+      const gsize stack_size = 0x8000;
+      g_autofree void *stack = g_malloc (stack_size);
+
+      /* Block signals */
+      sigset_t oldset, newset;
+      sigfillset (&newset);
+      sigprocmask (SIG_SETMASK, &newset, &oldset);
+
+      int pid = clone (invisible_helper, (char*)stack + stack_size, CLONE_VM | CLONE_NEWNS | SIGCHLD, NULL);
```

Comment:
I am fine with this as is, but I am wondering if it's really worth using the low level `clone()` API here instead of just invoking `fork() + setns()` ?

Hmm you know we could potentially use a `safe_fork()` helper like systemd has which also wraps the signal handling stuff. Anyways not a blocker, this code looks right to me as is.

Reply:
> I am fine with this as is, but I am wondering if it's really worth using the low level `clone()` API here instead of just invoking `fork() + setns()` ?

No.

Please take a look at the following lines:
```C
/* Bind mount the private mount namespace */
g_autofree char *ns = g_strdup_printf ("/proc/%d/ns/mnt", pid);
if (mount (ns, sysroot_ns, NULL, MS_BIND | MS_SILENT, NULL) < 0)
    err (EXIT_FAILURE, "failed to bind-mount sysroot-ns");
```

After spawning the child process, the parent process immediately bind mount its mount namespace to `sysroot_ns`. If we use `fork() + setns()`, the bind mounting may be executed after `fork()` but before `setns()`, which is a race condition. By using `clone()`, it is guaranteed that the child process is created with a new mount namespace and we are always bind mounting the new ns.

Reply:
I added some comments in code to explain it.